robot_name/robot_name.rb

```diff
@@ -2,31 +2,48 @@ class NameCollisionError < RuntimeError; end
 
 class Robot
   attr_accessor :name
-
-  @@registry
+  @@registry = []
 
   def initialize(args = {})
-    @@registry ||= []
-    @name_generator = args[:name_generator]
-
-    if @name_generator
-      @name = @name_generator.call
-    else
-      generate_char = -> { ('A'..'Z').to_a.sample }
-      generate_num = -> { rand(10) }
+    @name = args[:name] || generate_name
 
-      @name = "#{generate_char.call}#{generate_char.call}#{generate_num.call}#{generate_num.call}#{generate_num.call}"
+    if format_valid?(@name) && name_unique?(@name)
+      @@registry << @name
     end
+  end
 
-    raise NameCollisionError, 'There was a problem generating the robot name!' if !(name =~ /[[:alpha:]]{2}[[:digit:]]{3}/) || @@registry.include?(name)
-    @@registry << @name
+  def self.registry
+    @@registry
   end
+
+  private
+    def letter_randomizer
+      ("A".."Z").to_a.sample
+    end
+
+    def generate_name
+      name = []
+      2.times { name << letter_randomizer }
+      3.times { name << rand(10) }
+      name.join
+    end
+
+    def format_valid?(name)
```

Comment:
These look like boolean validators but are actually exception generators.

Conventionally, methods with `?` should simply return a boolean while exceptional code are often marked with `!`. Try to not mix the two behaviors.